### PR TITLE
added shoppingcart and cartitem entities anf functionality for them

### PR DESCRIPTION
**Pull request by polikarpov13**

### CODE REVIEW

File: src/main/java/book/store/onlinebookstore/model/ShoppingCart.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    private Long id;
+
+    @OneToOne
+    @JoinColumn(name = "user_id", referencedColumnName = "id")
+    private User user;
+
+    @OneToMany(mappedBy = "shoppingCart")
+    private Set<CartItem> cartItems;
+
+    @Column(nullable = false)
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/model/ShoppingCart.java:34]:
```suggestion
    @Column(nullable = false, name= name = "is_deleted")

```
cause hibernate want to match this value with isDeleted column

File: src/main/java/book/store/onlinebookstore/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    private ShoppingCart shoppingCart;
+
+    @ManyToOne
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+
+    @Column(name = "quantity", nullable = false)
+    private int quantity;
+
+    @Column(nullable = false)
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/model/CartItem.java:36]: same here

File: src/main/java/book/store/onlinebookstore/model/ShoppingCart.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import java.util.Set;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
+@Entity
+@SQLDelete(sql = "UPDATE sopping_carts set is_deleted = TRUE WHERE id = ?")
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/model/ShoppingCart.java:19]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_carts set is_deleted = TRUE WHERE id = ?")

```
be more attentive

File: src/main/java/book/store/onlinebookstore/dto/cartitem/CartItemDto.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    private Long id;
+
+    private Long bookId;
+
+    @NotNull
+    private String bookTitle;
+
+    @NotNull
+    @Min(1)
+    private int quantity;
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/dto/cartitem/CartItemDto.java]: remove empty lines

File: src/main/java/book/store/onlinebookstore/dto/cartitem/CartItemDto.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+public class CartItemDto {
+    private Long id;
+
+    private Long bookId;
+
+    @NotNull
+    private String bookTitle;
+
+    @NotNull
+    @Min(1)
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/dto/cartitem/CartItemDto.java]: its not necessary here
```suggestion
```

File: src/main/java/book/store/onlinebookstore/dto/shoppingcart/PrintShoppingCartDto.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+@Data
+public class PrintShoppingCartDto {
+    private Long id;
+
+    @NotNull
+    private Long userId;
+
+    @NotNull
+    private Set<CartItemDto> cartItemDtos;
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/dto/shoppingcart/PrintShoppingCartDto.java]: remove empty lines

File: src/main/java/book/store/onlinebookstore/dto/shoppingcart/ShoppingCartDto.java
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/dto/shoppingcart/ShoppingCartDto.java:1]: I think you can delete it and use only printShoppingCartDto (but rename it to ShoppingCartDto) but its up to you

File: src/main/java/book/store/onlinebookstore/dto/shoppingcart/ShoppingCartDto.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+@Data
+public class ShoppingCartDto {
+    private Long id;
+
+    @NotNull
+    private Long userId;
+
+    @NotNull
+    private Set<Long> cartItemIds;
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/dto/shoppingcart/ShoppingCartDto.java]: empty lines

File: src/main/java/book/store/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        }).collect(Collectors.toSet());
+    }
+
+    default Set<CartItem> mapModel(Set<CartItemDto> items) {
+        return items.stream().map(i -> {
+            CartItem item = new CartItem();
+            item.setId(i.getId());
+            return item;
+        }).collect(Collectors.toSet());
+    }
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/mapper/ShoppingCartMapper.java]: Why do you need all of those methods? Methods above should work correct

File: src/main/java/book/store/onlinebookstore/repository/shoppingcart/ShoppingCartRepository.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import book.store.onlinebookstore.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+
+    @Query("SELECT sc FROM ShoppingCart sc "
+            + "LEFT JOIN FETCH sc.cartItems ci "
+            + "LEFT JOIN FETCH ci.book b "
+            + "WHERE sc.user.id = (:id)")
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/repository/shoppingcart/ShoppingCartRepository.java]:
```suggestion
    @Query("SELECT sc FROM ShoppingCart sc "
            + "LEFT JOIN FETCH sc.user u "
            + "LEFT JOIN FETCH sc.cartItems ci "
            + "LEFT JOIN FETCH ci.book b "
            + "WHERE u.id = (:id)")

```

File: src/main/java/book/store/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public PrintShoppingCartDto getShoppingCart() {
+        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        return shoppingCartMapper.toPrintDto(
+                shoppingCartRepository.getShoppingCartByUserId(user.getId()));
+    }
```
Comment (Vasul-Pankiv) [inline src/main/java/book/store/onlinebookstore/service/shoppingcart/ShoppingCartServiceImpl.java:48]: you need to create another private method here to return a shoppingCart model and reuse this mathod into addItemToCart method, cause mapping from model to dto and then again to dto is not good

File: src/main/resources/db/changelog/changes/13-create-shopping-cart-table.yaml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+                  name: user_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (Vasul-Pankiv) [inline src/main/resources/db/changelog/changes/13-create-shopping-cart-table.yaml]: add empty lines to all your .yaml files

**Review by Vasul-Pankiv — CHANGES_REQUESTED** [review]

File: src/main/java/book/store/onlinebookstore/mapper/ShoppingCartMapper.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
```
Comment (olekskov) [inline src/main/java/book/store/onlinebookstore/mapper/ShoppingCartMapper.java]: These methods should be covered by relevant model mappers. In our case it's CartItemMapper and UserMapper.
Please check how to use one mapper in another

**Review by olekskov — CHANGES_REQUESTED** [review]

**Review by olekskov — APPROVED** [review]

**Pull request merged**